client/tests/integration/smartcontracts/multisig/src/lib.rs

```diff
@@ -0,0 +1,123 @@
+    Instruction(InstructionBox),
+    /// Accept vote for certain instruction
+    Vote(HashOf<InstructionBox>),
+}
+
+#[iroha_trigger::main]
+fn main(id: TriggerId, _owner: AccountId, event: EventBox) {
+    let (args, signatory): (Args, AccountId) = match event {
+        EventBox::ExecuteTrigger(event) => (
+            event
+                .args()
+                .dbg_expect("trigger expect args")
+                .try_into_any()
+                .dbg_expect("failed to parse arguments"),
+            event.authority().clone(),
+        ),
+        _ => dbg_panic("only work as by call trigger"),
+    };
+
+    let isi_hash = match &args {
+        Args::Instruction(isi) => HashOf::new(isi),
+        Args::Vote(isi_hash) => *isi_hash,
+    };
+    let votes_metadata_key: Name = format!("{isi_hash}/votes").parse().unwrap();
+    let isi_metadata_key: Name = format!("{isi_hash}/isi").parse().unwrap();
+
+    let (votes, isi) = match args {
+        Args::Instruction(isi) => {
+            FindTriggerKeyValueByIdAndKey::new(id.clone(), votes_metadata_key.clone())
+                .execute()
+                .expect_err("isi is already submitted");
+
+            let votes = BTreeSet::from([signatory.clone()]);
+
+            SetKeyValue::trigger(id.clone(), isi_metadata_key.clone(), JsonString::new(&isi))
+                .execute()
+                .dbg_unwrap();
+
+            SetKeyValue::trigger(
+                id.clone(),
+                votes_metadata_key.clone(),
+                JsonString::new(&votes),
+            )
+            .execute()
+            .dbg_unwrap();
+
+            (votes, isi)
+        }
+        Args::Vote(_isi_hash) => {
+            let mut votes: BTreeSet<AccountId> =
+                FindTriggerKeyValueByIdAndKey::new(id.clone(), votes_metadata_key.clone())
+                    .execute()
+                    .dbg_expect("isi should be submitted first")
+                    .into_inner()
+                    .try_into_any()
+                    .dbg_unwrap();
+
+            votes.insert(signatory.clone());
+
+            SetKeyValue::trigger(
+                id.clone(),
+                votes_metadata_key.clone(),
+                JsonString::new(&votes),
+            )
+            .execute()
+            .dbg_unwrap();
+
+            let isi: InstructionBox =
+                FindTriggerKeyValueByIdAndKey::new(id.clone(), isi_metadata_key.clone())
+                    .execute()
+                    .dbg_unwrap()
+                    .into_inner()
+                    .try_into_any()
+                    .dbg_unwrap();
+
+            (votes, isi)
+        }
+    };
+
+    let signatories: BTreeSet<AccountId> =
+        FindTriggerKeyValueByIdAndKey::new(id.clone(), "signatories".parse().unwrap())
+            .execute()
+            .dbg_unwrap()
+            .into_inner()
+            .try_into_any()
+            .dbg_unwrap();
+
+    // Require N of N signatures
+    if votes.is_superset(&signatories) {
```

Comment:
It's not hard to modify this code to work for `m` of `n` case. 
It would be some trigger metadata which would contain voting mode like `All`, `Majority`, `SuperMajority`, ...

Another way i was think about is to use this multisig trigger as some kind of gateway. 
So that signatories of multisig account can register another trigger to perform voting (in a different way) and then unregistered this trigger if no longer required.

Reply:
If you are referring to a dedicated ephemeral trigger corresponding to each multisig transaction, then yes, that would be an alternative. But so far the current design seems to be achieving an equivalent

Reply:
As far as the m of n feature is concerned,
one solution would be to expand "authentication state trie" during multisig proposals so that each node is `(quorum, is_pass)` and update the trie for each vote. This would also support recursive multisig:

```mermaid
flowchart
  classDef fail fill:#8c4351,fill-opacity:0.5
  classDef pass fill:#166775,fill-opacity:0.5

  012345678(1/2):::pass -->|012345| 012345(all):::fail
  012345678 -->|678| 678(2/3):::pass
  012345 -->|012| 012(2/3):::fail
  012345 -->|345| 345(2/3):::pass
  012 -->|0| 0( ):::pass
  012 -->|1| 1( ):::fail
  012 -->|2| 2( ):::fail
  345 -->|3| 3( ):::pass
  345 -->|4| 4( ):::pass
  345 -->|5| 5( ):::fail
  678 -->|6| 6( ):::pass
  678 -->|7| 7( ):::fail
  678 -->|8| 8( ):::pass

```

Reply:
> If you are referring to a dedicated ephemeral trigger corresponding to each multisig transaction, then yes, that would be an alternative.

I probably explained my point not precisely.

If multisig account signatories want to change voting scheme they can do the following:
1) multisig account `A` has multisig manager trigger `T1` which requires `N of N` votes
2) signatories of `A` make proposal and vote to register trigger `T2` which is also could be used to manage multisig account but requires `M of N` votes (or other voting scheme)
3) in the same proposal they also submit isi to unregister `T1` (and some other technical stuff like giving permission for signatories to execute `T2`)
 
 So now multisig account `A` has replaced voting scheme.

Reply:
> authentication state trie

Can you please explain a bit more about this trie? 

Is this would be a trie of proposal hashes?
So when new proposal is submitted we add a new leaf is added to the trie?

Reply:
Let me explain:

- the entire trie represents the readiness of one multisig transaction
- accounts are keys to nodes
  - `0` .. `8` are personal and the rest are multisig accounts
- node indicates whether the multisig transaction has been approved up to the corresponding account

Reply:
If we take this approach, we need a way to know its members from the multisig account in order to expand the trie